### DISCUSSION TITLE
How to override a models appearance using overrideModelAppearance

### DISCUSSION THREAD

**Discussion by candcconsulting** [iTwin/itwinjs-core#2527] [Q&A] [answered]
I am trying to override a models appearance with overrideModelAppearance.
Code is here
https://www.itwinjs.org/sandboxes/GlenWorrall_async/overrideModels
Am I interpreting the FeatureAppearance incorrectly ?
https://www.itwinjs.org/reference/imodeljs-common/rendering/featureappearance/
ie ColorDef.from
I did see this
https://github.com/imodeljs/imodeljs/blob/007be59f93813bc5525b1fe2afc48f20761bca56/core/common/src/test/FeatureSymbology.test.ts
Would overrideModel be better to use here ?

**Comment by pmconne** [comment]
overrideModelAppearance is the correct way to do this. The first time you asked this question you said:

> the FeatureAppearance documentation shows the fromRgb as using a ColorDef.from, but when I try using that in the call it reports various warnings.

What warnings?

**Reply by candcconsulting** [reply]
sorry that is why I put the sandbox so you could see how I was getting the ColorDef ...
          const colorValue =  ColorDef.fromString(colourList[i]);
          IModelApp.viewManager.selectedView!.overrideModelAppearance(row.id, FeatureAppearance.fromRgb(ColorDef.fromString(colourList[i])));                  

I just put ColorDef.from back in the same line .. but generally I had colorValue there to validate if that is what is intended
          IModelApp.viewManager.selectedView!.overrideModelAppearance(row.id, FeatureAppearance.fromRgb(colorValue));      
however I set the color, the override does not seem to be applied

**Comment by pmconne** [comment]
Something appears to be wrong with the sandbox app, not your code. See [my fork](https://www.itwinjs.org/sandboxes/pmconne/initial-peach-koi-373) of the sandbox you linked.
Some things to note:
- In Api.tsx I added a line that attempts to output the Ids of all the models in the view. It produces an exception. Most of the time, with that line in place, the view will start loading and displaying the models in their normal colors, and then suddenly start displaying them as red, green, blue, and purple as specified by my overrides.
- If I comment out that line, it almost never displays the models using my color overrides.
- In App.tsx, I override the ViewFlags to turn off background map and set render mode to wireframe. This never has any effect.

@Josh-Schifter @williamkbentley @BaoTon 

EDIT: I just ran it again and it began displaying the models in normal colors in shaded mode, then displayed them in my override colors, then switched to displaying them in wireframe using their normal colors.

**Reply by BaoTon** [reply]
@StefanRetief, are modules loading properly for sandbox in this case?

**Reply by candcconsulting** [reply]
Thanks Paul.  FYI In the "app" it always displays all models in grey, so hence I thought I was using the call incorrectly.
I can add you to the github repo but it is using some special data from the iModel which I think you will not have access to.

**Comment by pmconne** [comment, marked as answer]
@candcconsulting To set up your initial view, supply an async function that accepts an IModelConnection and returns a Promise<ViewState> as part of your viewportOptions, instead of doing so in an onIModelConnected callback.
I've updated [my fork of your sandbox](https://www.itwinjs.org/sandboxes/pmconne/initial-peach-koi-373) to illustrate this. It displays each model as one of six different colors.

**Reply by candcconsulting** [reply]
Thanks Paul ... I had a "few" things wrong .. .but managed to work it out with your snippet ... will plug it into my main app.
